pkg/controller/workload/workload_stream_test.go

```diff
@@ -0,0 +1,301 @@
+	return s
+}
+
+func (f *MockDiscovery) StreamAggregatedResources(server discoveryv3.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error {
+	numberOfSends := 0
+	for {
+		select {
+		case <-f.close:
+			return nil
+		case resp := <-f.responses:
+			numberOfSends++
+			log.Infof("sending response from mock: %v", numberOfSends)
+			if err := server.Send(resp); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (f *MockDiscovery) DeltaAggregatedResources(server discoveryv3.AggregatedDiscoveryService_DeltaAggregatedResourcesServer) error {
+	numberOfSends := 0
+	for {
+		select {
+		case <-f.close:
+			return nil
+		case resp := <-f.deltaResponses:
+			numberOfSends++
+			log.Infof("sending delta response from mock: %v", numberOfSends)
+			if err := server.Send(resp); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func TestWorklaodStreamCreateAndSend(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+	workloadStream := WorkloadStream{
+		Event: nil,
+	}
+
+	// create a fake grpc service client
+	mockDiscovery := NewMockServer(t)
+	conn, err := grpc.Dial("buffcon",
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithBlock(),
+		grpc.WithContextDialer(func(context.Context, string) (net.Conn, error) {
+			return mockDiscovery.Listener.Dial()
+		}))
+	if err != nil {
+		t.Errorf("grpc connection client create failed, %s", err)
+	}
+	defer conn.Close()
+	client := discoveryv3.NewAggregatedDiscoveryServiceClient(conn)
+	stream, streamErr := client.DeltaAggregatedResources(ctx)
+	if streamErr != nil {
+		t.Errorf("create stream failed, %s", streamErr)
+	}
+	workloadStream.Stream = stream
+
+	patches1 := gomonkey.NewPatches()
+	patches2 := gomonkey.NewPatches()
+	tests := []struct {
+		name       string
+		beforeFunc func()
+		afterFunc  func()
+		wantErr    bool
+	}{
+		{
+			name: "test1: send request failed, should return error",
+			beforeFunc: func() {
+				patches1.ApplyMethod(reflect.TypeOf(client), "DeltaAggregatedResources",
+					func(_ discoveryv3.AggregatedDiscoveryServiceClient, ctx context.Context, opts ...grpc.CallOption) (discoveryv3.AggregatedDiscoveryService_DeltaAggregatedResourcesClient, error) {
+						return stream, nil
+					})
+				patches2.ApplyMethod(reflect.TypeOf(workloadStream.Stream), "Send",
+					func(_ discoveryv3.AggregatedDiscoveryService_DeltaAggregatedResourcesClient, req *discoveryv3.DeltaDiscoveryRequest) error {
+						return errors.New("timeout")
+					})
+			},
+			afterFunc: func() {
+				patches1.Reset()
+				patches2.Reset()
+			},
+			wantErr: true,
+		},
+		{
+			name:       "test2: no movk, send request successful, should return nil",
+			beforeFunc: func() {},
+			afterFunc:  func() {},
+			wantErr:    false,
+		},
+		{
+			name: "test3: fail to create workloadStream, should return error",
+			beforeFunc: func() {
+				patches1.ApplyMethod(reflect.TypeOf(client), "DeltaAggregatedResources",
```

Comment:
I think you donot need to patch here, you can just put this resource into `f.responses`.

Reply:
This test case is to cover the case when creating a stream fails. How can I use `f.responses` to cause stream creation to fail?

Reply:
If create stream failed, mostly because of auth/network
But this test does not include auth or network